src/ndarray/ndarray.cc

```diff
@@ -558,6 +558,101 @@ void CopyFromTo(const NDArray& from, const NDArray& to, int priority) {
   }
 }
 
+void Quantize(const NDArray &from, NDArray *to, NDArray *residual, const std::string& compress,
+              const float neg_threshold, const float pos_threshold,
+              int priority) {
+  CHECK(from.shape().ndim() != 0)
+      << "source operands have zero dimension shape";
+  // important: callback must always capture by value
+  NDArray ret = *to;
+  NDArray res = *residual;
+  int a = from.ctx().dev_mask();
+  int b = to->ctx().dev_mask();
+  if (a == cpu::kDevMask && b == cpu::kDevMask) {
+    if (compress == "2bit") {
+      Engine::Get()->PushSync([from, res, ret, neg_threshold, pos_threshold](RunContext ctx) {
```

Comment:
res and ret passing by value is expensive. Please pass as reference or pass the pointers

Reply:
Is there a reason why some functions in ndarray.cc make a copy of NDArray from a pointer and pass the copy like here? I wasn't sure if there was a reason, so I did the same. So it is safe to pass by reference into this function and pass it? I'll change this

Reply:
That's a combination of a couple of bad things:
1) lint complains about using a non-const reference as a parameter, and prefers non-const pointer.  In that case, some ppl just make it a const reference and then deal with that later (as you are seeing).
2) In an attempt to satisfy this lint error, some ppl have made the "destination" or "modified parameter" a non-const pointer.
3) Since most NDArrays are passed as 'const reference', in order to get the non-const pointer, a copy has to be made.  This is a hack to satisfy the two problems above.
4) Current thinking is to nto use the non-const pointer thing as you can see from changes such as this:  (in init_op.h): https://github.com/apache/incubator-mxnet/commit/1637011c2a9e168feefafe60573bb2181c3a5d19
5) While this sort of thing is hacked here and there, it is not desired -- especially for new code.

They're trying to get rid of the 'const'